contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+
+        //get token
+        IOptionToken token = IOptionToken(tokenAddress);
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        cost = optionViews.getBuyOptionCost(market, optionType.isLongToken, optionType.strikeIndex, purchaseAmount);
+    }
+
+    /**
+     * @notice Amount of profit made from exercising an option contract (current price - strike price). 0 if exercising out-the-money.
+     * @param options is the address of the options contract
+     * @param amount is the amount of tokens or options contract to exercise. Only relevant for fungle protocols like Opyn
+     */
+    function exerciseProfit(
+        address options,
+        uint256,
+        uint256 amount
+    ) public view override returns (uint256 profit) {
+      //get token
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      // Get strike index, whether is long
+      OptionType memory optionType = addressToOptionType[options];
+
+      //profit of exercising
+      profit = optionViews.getSellOptionCost(market, optionType.isLongToken, optionType.strikeIndex, amount);
+    }
+
+    function canExercise(
+        address options,
+        uint256 optionID,
+        uint256 amount
+    ) public view override returns (bool) {
+        //get token
+        IOptionToken token = IOptionToken(options);
+        //get market
+        IOptionMarket market = IOptionMarket(token.market());
+
+        if (block.timestamp < market.expiryTime()) {
+            return false;
+        }
+        if (exerciseProfit(options, optionID, amount) > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @notice Purchases the options contract.
+     * @param optionTerms is the terms of the option contract
+     * @param amount is the purchase amount in Wad units (10**18)
+     * @param maxCost is the max amount of paymentToken to be paid for the option (to avoid sandwich attacks, ...)
+     */
+    function purchase(
+        ProtocolAdapterTypes.OptionTerms calldata optionTerms,
+        uint256 amount,
+        uint256 maxCost
+    ) external payable override returns (uint256) {
+      require(
+          block.timestamp < optionTerms.expiry,
+          "Cannot purchase after expiry"
+      );
+
+      //get token address
+      address tokenAddress = lookupOToken(optionTerms);
+      if(tokenAddress == address(0)){
+        populateOTokenMappings(optionTerms);
+      }
+
+      tokenAddress = lookupOToken(optionTerms);
+
+      require(tokenAddress != address(0), "Market needs to exist!");
+
+      // Get strike index, whether is long
+      OptionType memory optionType = addressToOptionType[tokenAddress];
+      //get token
+      IOptionToken token = IOptionToken(tokenAddress);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+      IERC20 baseToken = market.baseToken();
```

Comment:
I need some context here, why do we have to invoke `baseToken` if we already know the `optionTerms.underlying` from the parameters?

If we can save gas here by using `optionTerms.underlying`, we should.

Reply:
yep, good spot